### ISSUE TITLE
to be or not to be: CMake install() 

### ISSUE THREAD

**Issue by adevress** [abseil/abseil-cpp#38]
Issue following the discussions on #8 : should we support install() in the cmake build:

Pro:
- required by people who do not use cmake in their project.
- make the usage of package manager possible

Cons:
- make easy for to rely on ABI and have problems

**Comment by rogeeff** [comment]
Another con:
  bazel BUILD files will not have install rules. And we want the two build files to be almost identical. In fact I think we'll end up auto generating CMakefile(s) out of corresponding bazel files once dust is settled.

We can consider separate CMakefile, which is going to be provided as a courtesy and only present some install rules. This CMakefile will not be generated and can get out of sync.

**Comment by lilianmoraru** [comment]
Without the install step you can't use it in SDKs and Linux Distros.
It will be just a bunch of projects, everybody using their own version of Abseil(most probably, locked at one commit for the rest of the project's life)...

**Comment by adevress** [comment]
@rogeeff  Bazel has a very specific workflow where everything is built inside a single workspace without any installation. This is not the case for almost every other built system and packaging system existing in this world that follow a configure / make / make install pattern. 
I do not think it is a reasonable option to not be compatible with them, just to simplify Bazel build file maintainability.

Bazel might be omipresent at Google, but it is not the case (yet) in the open source world, and will probably never be the case some professional environment.

If maintainability of Bazel build file is an issue, why not "annotate" the target that are installable in the bazel recipe and generate the CMake Files accordingly ?

**Comment by Sarcasm** [comment]
Another pro:

"workspace build system", based on CMake or not (often not), often rely on the install output.
They offer way to copy files manually, but this is useful mostly when a project lacks an install  target. The dependency between project is *based on* the install step.

Some tools I know of:
- [bitbake](https://www.yoctoproject.org/docs/1.6/bitbake-user-manual/bitbake-user-manual.html)
- CMake external projects (by default)
- catkin and ament from the [ROS](http://www.ros.org/) project
- [vcpkg](https://github.com/Microsoft/vcpkg)
- [cget](https://github.com/pfultz2/cget)
- [nix](https://github.com/NixOS)
- [conan](https://www.conan.io/)
- [hunter](https://github.com/ruslo/hunter)
- [buildroot](https://buildroot.org/) (some companies have forks of this or similar things)

I don't know of such tools that don't depend on install actually.

I'm wondering if there any CMake user that does not want the install target and feel strongly about it?

**Comment by adevress** [comment]
@Sarcasm  :+1:  Yep, and you can add spack to the list for HPC environments

**Comment by lilianmoraru** [comment]
@Sarcasm Add to that Distros package managers(RPM for example, where it has a "buildroot" and takes the files from there after you run "make install" into it).

**Comment by a4z** [comment]
I would like to follow the advice to live on head, have only workspaces and source builds, but in reality I have projects where I need go back in time.
When I need to fix a 5 year old version of a software, I am not allowed to change any part that is not directly affected from the problem.
The way to ensure this is, for every component the system is build with:
`make install DESTDIR=/where/ever`
pack the result, sing it and put a check sum int some archive.
The fixed version will have the same binary base system as the version 5 years ago, guaranteed.

Also, I need to deliver exactly a known binary package of today software to developers for today development. The way to ensure this is, the same.

If I can not do this I can not use a library at work, I am not allowed to.

Now, it is not that hard, just annoying, to write an install script yourself, but, 
when I have to fear that after each update of an library my self made install script is incomplete/breaks, practically speaking a library will not be used. 

And I would like to use abseil, because todays development should be on head, and the description for this lib and the concept behind sound just too good and fits exactly some of my needs.
But I have to live with both worlds, and I try to build a bridge. And therefore, I would need a `make install DESTDIR..` because that's what is used to solve some practical problems I have to live with.

Could I please have it?

**Comment by lilianmoraru** [comment]
From the previous issue, it seems that people assume that make install means precompiled binaries by default.
It doesn’t need to be that way, you can install the source if you want and find_package can even do weird things like adding the project as a subdirectory(would not recommend)...

**Comment by gennadiycivil** [comment]
In the spirit of [the documentation](https://abseil.io/about/compatibility) - it clearly says "Do not depend on a compiled representation of Abseil. We do not promise any ABI compatibility — we intend for Abseil to be built from source, hopefully from head"
So.. if people would like to include install() that is fine of course , however it clearly goes against the spirit here and therefore should not be a part of master.

**Comment by Sarcasm** [comment]
> In the spirit of the documentation - it clearly says "Do not depend on a compiled representation of Abseil. We do not promise any ABI compatibility — we intend for Abseil to be built from source, hopefully from head"

Please, consider one moment that most persons in this issue understand that.

> So.. if people would like to include install() that is fine of course , however it clearly goes against the spirit here and therefore should not be a part of master.

It's not so black and white:

- `install()` does not mean binary distribution
- people who build from source, often do so by using the install targets


The build tools I mentioned earlier, https://github.com/abseil/abseil-cpp/issues/38#issuecomment-338779633, and [spack](https://github.com/LLNL/spack) mentioned by @adevress, https://github.com/abseil/abseil-cpp/issues/38#issuecomment-338785025, all relies on `install()`.

And for the subset I actually used: `bitbake`, `buildroot`, `buildroot-like`, `catkin`, `ament`,
and by the look of it, `nix` and `spack` too.
All of these, build workspaces from sources.

They all respect this contract:

> In your project build, ensure there is exactly one (current) copy of Abseil.
> All of the dependencies that require Abseil will work on that. Build it from source.
> -- @tituswinters: https://github.com/abseil/abseil-cpp/pull/8#issuecomment-338191755

**Comment by gennadiycivil** [comment]
Interesting argument @Sarcasm, thank you. 
> install() does not mean binary distribution

I tend to think about install() as a binary distribution, copy the build artifacts within the file systems accessible from where this build is running. 

It would perhaps be useful to consider this from a different point of view, could you please elaborate, how you see what install() does?

Thank you again.
G

**Comment by a4z** [comment]
Interesting quote from tituswinters, @Sarcasm , and true what you write
For some, they way to ensure that there is only one copy of a component is that developers have only access to one version distributed via a package manager -or as part of the SDK- used for the current product development. 
The package hast to be generated via `make install DESTDIR`, of course.

In terms of ABI compatibility , since it is normal to rebuild always everything that depends on top a component when a component is updated, this is not an issue at all. (as long as not the MS way of doing things is chosen to break ABI between Debug/Release builds. Abseil is not planning to do this?)

Seriously, in some working environment, when you say every workspace (team) needs to have a source version from Abseil, this is the guarantee for non equal Abseil version used in a product  because some team will update today, others tomorrow ....
But when all have to use the same SDK, they will have the same version guaranteed.

I find it very interesting that there are 2 such different point of views and requirements to reach same target, what is, the same version, hopefully the latest, of a component in a product development environment.

Well, its up to the repo owner/developer to decide, I provided a different point of view, everything else could only be repeat my self, so my 2 cents are donated with this and my previous post. ;-)

**Comment by adevress** [comment]
> I tend to think about install() as a binary distribution, copy the build artifacts within the file systems accessible from where this build is running.

@gennadiycivil That's the source of the mis-understanding I think.

For many package manager and build system, install() is just an intermediate step in the build pipeline what can be understood as "please, put the public interface of component X of revision Y under path Z for reuse by the next component in the build".

For package manager like Nix or Spack, install() of a intermediate component (like abseil) is just a step in the build process before continuing the build for the other component that depend on it. 

Nix detects the changes in both the revision and the build configuration, and determine a unique, immutable install() path based on a cryptographic hash(recipe, source, dependencies) where the component (abseil) will be installed. 

 This has several advantages:
- You definitively solve all ABI problems, because a change in the abseil source will generate a different output, under a different prefix, and trigger recompilation of every component depending on it.
- You have a very clear separation between every component: each component can have its own build system ( cmake, autottools, meson, scons ), as long as they all provider an "install under prefix" command, everything interoperate nicely.
- Every tuple <component, configuration, dependencies> is built once and only once. Avoiding useless recompilation.

Spack and some others works the same way.

It seems to me that you guys are afraid of install() phase because you suffered in the past of the limitations of classical Linux distro package manager (rpm, dpkg).
With RPM/DPKG,  every component is shipped into a global namespace, under the same prefix and relying on strong assumption on ABI and API compatibility. 
The consequence is that it triggers happily nuclear explosion each time a non-ABI-compatible update (of googletest) is deployed. 
And the second consequence (side effect) of that, is that everybody use outdated package because they can not get updated without explosion.

This is not what many of us rely on.

**Comment by lilianmoraru** [comment]
@adevress It's not an issue with RPM/DPKG per se but more with how Linux distros traditionally work.
We have at work projects that use BitBake but also RPM for dependency management(so we don't have to reinvent a build-system and port everything to it).
Example of RPM-based project that lives at HEAD:
```
new version of Abseil ---- project1 ------------ project2 ---------- etc...
            \ install         / use abseil          / use abseil
SDK -------------------------------------------------------------------------------                                           
                                         [abseil-HEAD]
```

After each merged commit, the SDK is updated in Git(LFS) and when somebody does `repo sync -j $(nproc)`, he gets the latest SDK(with the latest interface, libraries, etc... modifications) and his code update.

A team can basically have just an SDK that is always at HEAD and only the source code of their own project(a team can have their own manifest, the CI needs to import them all with its own manifest, make the build and update the SDK with libraries, headers, RPM database, etc...).

**Comment by adevress** [comment]
@lilianmoraru  I totally agree. The problem is not so much RPM / DEB itself, more how it is traditionally used.

**Comment by Sarcasm** [comment]
@gennadiycivil

I have tried to produce a "simple" CMake file that ressemble what some of the mentioned tools do.
I picked a method that uses CMake external projects, not because this is the best tool, but because the example is self contained and hopefully most people have CMake installed somewhere.

The examples build cctz and its dependencies, Google Mock and Google/benchmark.
All of the sources are fetched from the master branch on Github.
None of these projects "include" the other, which is convenient because it's possible to reuse them.

First create the project:

```
mkdir ep-cctz
cd ep-cctz
cat <<'EOF' > CMakeLists.txt
cmake_minimum_required(VERSION 2.8.12)

include(ExternalProject)

set_property(DIRECTORY PROPERTY EP_BASE ep)

ExternalProject_Add(ep_benchmark
  URL https://github.com/google/benchmark/archive/master.zip
  CMAKE_ARGS -DBENCHMARK_ENABLE_TESTING=OFF -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>)
ExternalProject_Get_Property(ep_benchmark install_dir)
set(ep_benchmark_install_dir "${install_dir}")

ExternalProject_Add(ep_gmock
  URL https://github.com/google/googletest/archive/master.zip
  CMAKE_ARGS
      -DBUILD_GMOCK=ON
      -Dgmock_build_tests=OFF
      -Dgtest_build_samples=OFF
      -Dgtest_build_tests=OFF
      -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>)
ExternalProject_Get_Property(ep_gmock install_dir)
set(ep_gmock_install_dir "${install_dir}")

ExternalProject_Add(ep_cctz
  URL https://github.com/Sarcasm/cctz/archive/master.zip
  LIST_SEPARATOR ::
  CMAKE_ARGS
      -DBUILD_EXAMPLES=ON
      -DBUILD_TESTING=ON
      "-DCMAKE_PREFIX_PATH=${ep_benchmark_install_dir}::${ep_gmock_install_dir}"
      -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
  DEPENDS ep_benchmark ep_gmock)

# # Here, for example, Abseil could reuse ep_cctz and ep_gmock to be built.
# # However the current proposed CMake, which IMPOSE cctz to be in a subdirectory,
# # does not work.
# # I really wish the CMake integration was "classic".
# # How can cctz be embedded and be used by other projects?
# # This works only if abseil is the only user of cctz,
# # this does not compose well.
# ExternalProject_Add(ep_abseil
#   # URL https://github.com/adevress/abseil-cpp/archive/master.zip
#   URL https://github.com/adevress/abseil-cpp/archive/b5ce9df6243286729bb5c23f848c92c43bcd9ff0.zip
#   LIST_SEPARATOR ::
#   CMAKE_ARGS
#       -DBUILD_EXAMPLES=ON
#       -DBUILD_TESTING=ON
#       "-DCMAKE_PREFIX_PATH=${ep_cctz_install_dir}::${ep_gmock_install_dir}"
#       -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
#   DEPENDS ep_cctz ep_gmock)
EOF
```

Then build it with CMake:

```
mkdir build
cd build
cmake ..
make
```

Now, to understand what is going on:
- we have a "workspace" description in the CMakeLists.txt
- we build projects from sources
- yet, we use the `install()` provided by these projects as a dependency between projects. Please note that, all other Google projects provide an install target, and this is nice of them, if only Abseil was doing the same.

If we take a look at the build layout, in the `build/ep/` directory, we have this:

```
../ep-cctz/build$ ls ep
Build  Download  Install  Source  Stamp  tmp
```

This is similar to some of the tools I mentioned.

Here is the content of the directories,
which matches the different stages of the build that most of these build tools uses.


- `Download/`: the directory where the sources are downloaded: `wget ...`

  ```
  $ tree ep/Download
  ep/Download
  ├── ep_benchmark
  │   └── master.zip
  ├── ep_cctz
  │   └── master.zip
  └── ep_gmock
      └── master.zip
  ```

- `Source/`: where they are extracted: `unzip ...`

  ```
  $ tree -L 2 ep/Source
  ep/Source
  ├── ep_benchmark
  │   ├── appveyor.yml
  │   ├── AUTHORS
  │   ├── cmake
  │   ├── CMakeLists.txt
  <snip...>
  │   ├── src
  │   ├── test
  │   └── tools
  ├── ep_cctz
  │   ├── BUILD
  │   ├── cmake
  │   ├── CMakeLists.txt
  <snip...>
  │   ├── README.md
  │   ├── src
  │   └── WORKSPACE
  └── ep_gmock
      ├── appveyor.yml
      ├── BUILD.bazel
      ├── CMakeLists.txt
      ├── googlemock
      ├── googletest
      ├── README.md
      ├── travis.sh
      └── WORKSPACE
  ```


- `Build/`: the build directory, using "out-of-source" build: `cmake .. && make ...`

  ```
  $ tree -C -L 2 ep/Build
  ep/Build
  ├── ep_benchmark
  │   ├── CMakeCache.txt
  │   ├── CMakeFiles
  │   ├── cmake_install.cmake
  │   ├── install_manifest.txt
  │   ├── Makefile
  │   └── src
  ├── ep_cctz
  <snip...>
  │   ├── install_manifest.txt
  │   ├── libcctz.a
  │   ├── Makefile
  │   ├── Testing
  │   ├── time_tool
  │   ├── time_zone_format_test
  │   └── time_zone_lookup_test
  └── ep_gmock
      ├── CMakeCache.txt
      ├── CMakeFiles
      ├── cmake_install.cmake
  <snip...>
      ├── install_manifest.txt
      └── Makefile

  ```

- `Install/`: packages are installed in the build tree, so that other built package can depend on it:
  `make install`

  ```
  $ tree -C -L 3 ep/Install
  ep/Install
  ├── ep_benchmark
  │   ├── include
  │   │   └── benchmark
  │   └── lib
  │       ├── cmake
  │       └── libbenchmark.a
  ├── ep_cctz
  │   ├── include
  │   │   └── cctz
  │   └── lib64
  │       ├── cmake
  │       └── libcctz.a
  └── ep_gmock
      ├── include
      │   ├── gmock
      │   └── gtest
      └── lib64
          ├── libgmock.a
          ├── libgmock_main.a
          ├── libgtest.a
          ├── libgtest_main.a
          └── pkgconfig
  ```


The build happens in sequence of dependencies:

- gmock and benchmark first
- then cctz which depends on them

Also, for each of these projects, the following stages can happen, in the following order:
- download
- extract
- patch
- configure
- build
- install

When a package like cctz depends on other projects, we point it to the earlier projects install tree,
so that it can find the headers and library.

You can see this for the `cctz` project, where we have:

```
ExternalProject_Add(ep_cctz
  <snip...>
  CMAKE_ARGS
      <snip...>
      "-DCMAKE_PREFIX_PATH=${ep_benchmark_install_dir}::${ep_gmock_install_dir}"
      <snip...>
  DEPENDS ep_benchmark ep_gmock)
```

This means `cctz` will look in the install tree of `benchmark` and `gmock`,
when calling the `find_package()` command from CMake:

```cmake
find_package(benchmark REQUIRED)
find_package(GMock REQUIRED)
```

-- https://github.com/Sarcasm/cctz/blob/44ddb7563e46e23ecd32a67896570b4b4d81edeb/CMakeLists.txt#L101-L102

CMake external projects is not the best example, tools like `nix`, `spack`,
`catkin`, `ament` or `bitbake` offer more control and scale better for big projects,
but this example demonstrates how projects can rely on each other `install` target
when building a workspace from source.

It also demonstrates that in the current state, without `install()` targets,
and by embedding it's dependencies in the CMakeLists.txt,
the Abseil CMake is not usable at all in these scenarios, which is very unfortunate IMHO.

**Comment by Orphis** [comment]
@Sarcasm This is potentially using different ABIs. If your top level project uses -std=c++14 and wants to propagate it, you will have to pass it manually to each dependency. This version of ExternalProject is broken in that regard, and you should have a look at the GoogleTest documentation for a better usage.

Abseil internally changes depending on what is available in the standard library. If you can have different standards used when you compile it and when you use it, you will have issues linking, or worse, difficult to debug crashes at runtime.

**Comment by Sarcasm** [comment]
It's true that this toy example is not perfect.
I did not mean to have a "top level" project here, I assumed the CMakeLists.txt was the whole, final workspace.

To be more correct, a toolchain file may be needed, e.g.:

```cmake
set(CMAKE_CXX_FLAGS "-std=c++14")
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)
```

the projects may need to be patched when they don't honor the toolchain, ...
All of these, the tools can do it, and often do.
Maybe not so often for the reasons of the Abseil compatibility requirements.


Actually, I'm wondering if the question of whether or not having an `install()` is the right question.
What if we assume that it is "yes", and that we want Abseil to be a good citizen in the CMake world.
What could be done to make Abseil safer in this scenario?
What if Abseil wasn't "Zero configuration" as advertised here:
- https://opensource.googleblog.com/2017/09/introducing-abseil-new-common-libraries.html

Without giving it too much thoughts, a configuration could reasonably be made at build time, and installed, to either:
- have a frozen configuration from Abseil build time
- or assert the configuration is compatible/identical

Since Abseil is not a header only, I personally don't see any strong benefits of being configuration-free.
I expect I may be proven wrong here.

I understand Bazel may not need configuration as there is only one possible toolchain and build settings at anytime, or something along these lines.
CMake is just not like that, but the project can surely be accommodated to avoid misuse of the library.

Also, I think this is a goal of modern CMake versions to support this kind of scenario, for example, looking at this recent discussion:
- https://gitlab.kitware.com/cmake/cmake/issues/17146#note_299405

Ideally people could specify the minimum standard required in a CMakeLists.txt, but the user can request a specific version from outside.

**Comment by ghost** [comment]
Just dropping in here (I don't have a stake in this library, but some experience with CMake).. The typical workflow with CMake is to build everything from source. Especially on windows, you need to download all of your dependencies and then build them. Everything uses the same compiler. On OSX, you have homebrew, which uses an ABI-compatible method of building (everything is build with the same compiler, with the same options).

I don't think  being obstinate about using install is going to fix any problems or make anyone more likely to use `HEAD`. It's just going to make other people have to write their own CMakeLists.txt to get around your obstinance.

People don't read the documentation and then complain. They may file issues when they shouldn't. The way that homebrew deals with that problem is to ban people who don't read the documentation. If you're that concerned about extra work, then you should take the same approach.

**Comment by junyer** [comment]
In order for RE2 to depend on Abseil without abandoning or alienating our external users, I believe that it will be necessary for Abseil to accommodate traditional use cases in some way. Perhaps the "long-term support" branches could serve as a basis for packaging? And then Debian could handle Abseil much like they handle Boost, I imagine, and likewise for other distributions.

**Comment by ghost** [comment]
>  Perhaps the "long-term support" branches could serve as a basis for packaging?

This is exactly what the maintainers are afraid of: people conflating `install` with long-term support for an outdated version. I completely understand this and I personally have little sympathy for users of outdated software. 

The problem is that linux distributions have a completely different packaging model than OSX or windows. In the latter two, you just use the latest version of the library, or something close to it. But in the case of linux, the packages are frozen in time for years.

One solution could be to only offer `install` if the OS is OSX or windows, making it intentionally a PITA to package abseil on linux distributions. That could offer all of the advantages and none of the disadvantages (well, except for linux packagers, which aren't going to be supported anyway).

**Comment by lilianmoraru** [comment]
@xoviat Now you've made it very clear why the `install` target is not desired...

I thought initially that the main argument against the `install` target was because of the ABI issues when compiling/linking libraries with different flags for example(I was trying to explain that it does not have to be precompiled, but it does not address the "love for old, buggy, slow code" issue).

I understand that the idea is to not leave developers any other option than to directly use abseil(because distros could not "make install" abseil, for example :) ), _potentially_ living at "HEAD".

Looking at other open-source projects, the reality is that projects usually pin the submodule or copy the code at one point in time and leave it there for "stability" or because nobody considers it important to have the code up-to-date(even when there are critical security fixes :( ).

What I'm trying to say is that this still won't ensure that the projects live at "HEAD"(but does improve the chance of it, because it might exclude support for abseil in Linux distros), it's just that it makes it considerably harder for the SDK, big projects(many smaller components/repos depending on abseil), etc... use-case to happen.
<br/>

I am not sure yet which solution is better or which issue is actually more important(did not yet take time to thoroughly think about this), but I can propose a somewhat weird compromise.

I don't remember which programming language had this(I thought Ruby, but can't find it) but there was a very specific problem in the language that you could not solve easily with the usual API, but also you could not apply the obvious "workaround" because it was awful(for performance or something like this).
In the end the language provided the solution, for that exceptional case where you need it, and it was called somewhat like this: `i_am_a_horrible_person_do_stuff`.

Maybe we can still have the `install` target but hidden under a cumbersome and obviously "bad" flag(which you will want to read about)?
Something like: `-DI_SHOULD_NOT_USE_THIS_AND_INCLUDE_ABSEIL_AS_A_SUBDIRECTORY_INSTEAD=TRUE`? :smile:

**Comment by junyer** [comment]
> This is exactly what the maintainers are afraid of: people conflating `install` with long-term support for an outdated version. I completely understand this and I personally have little sympathy for users of outdated software.

https://abseil.io/about/philosophy states:

> We will periodically mark a tag as “supported” and branch it — if we discover security issues or major performance problems, we’ll update those branches. Our expectation is to do this every 6 months and support those branches for 2 years.

I understand that Abseil folks don't wish to endorse or even encourage packaging, but considering their promise to maintain such branches, I don't understand why they would obstruct packaging.

There will be users of RE2, for example, who won't know or care about Abseil and its philosophy. In many cases, they will be using language bindings, so they won't know or care about C++ problems. In all cases, they will want new versions of RE2 to work just as easily as old versions of RE2. I can't condemn them to using outdated software simply because they had the temerity to use RE2 before Abseil existed.

**Comment by JonathanDCohen** [comment]
@xoviat can you please go on about the linux packaging model?  I thought apt-get largely worked similarly to homebrew et al.

**Comment by a4z** [comment]
some difference between Windows and Linux packaging:
(sorry for dropping in to this topic, but I feel it could not harm to add some info)

on Windows MSVC break ABI from version to version and even between debug and release build.

This is why in Windows project often are huge monoliths, with external dependency in a subdirectory, 3rdParty or so.
The software, if it is source code, becomes old over time also, or, some project have even binaries in such folders.
If they have only binaries and no source, than they can not update the compiler, also not uncommon in reality.

On Linux ABI does not break that often, I think I had it only once, some when during gcc 3.x
This is why it is not a problem to run multiple gcc versions, or also clang, and compile programs with them that use the same binary versions of libraries in the system.

on RHEL you use for example software collections to get a new gcc-63 or whatever, the latest python or ruby, .....
and these latest version work great together with existing components, build with a different compiler.
you add them to your system and the system itself stays stable as it is.

(I totally understand why abseil would break in such a scenario, I have no understanding why a lib like google test should not be shipped via a package, but this just as a site note)

Of course RHEL, Suse, or debian do not update basic packages to the latest version, they add latest version in parallel, this is for a reason.
If you want always the latest in your base system, you run Arch or Slackware current or something similar.
If you look at this distros, what xoviat wrote is not true.
But, these distros, and Debian, do not rebuild world for releases, they are rolling releases, they add and replace.
In contrast to RHEL or SUSE who can rebuild everything, if they want or mean it is required.
Or, when you build for embedded via yocto, where you also rebuild everything.

So there is no one story about package managers on Linux.

On Windows such a scenarios, any of them, are impossible with MSVC, therefore you need those monolithic workspaces with insane build times and everything included.

I have no idea about Mac.

**Comment by ghost** [comment]
> @xoviat can you please go on about the linux packaging model? I thought apt-get largely worked similarly to homebrew et al.

No, it doesn't. Because `apt-get` is integrated into the operating system, the software available is restricted to a curated subset that was present at the time of the OS release (for example, gcc 6 is not present by default in the Ubuntu repository). This is especially painful for something like RHEL, where the software is years old.

In contrast, homebrew usually stays a few months behind HEAD at most. Specifically, let's compare the effort to update a package in both package managers.

1. In something like Ubuntu, you need to go through a curation process, the result of which will be included in the next OS release. This can take at least a few weeks of testing and days of effort.
2. In homebrew, if you want to update the version, you just file a PR against `homebrew-core`. The CI automatically tests the latest version of the package and the PR is accepted within a week at most. In addition *homebrew forces users to use the latest version of the package*. See this [comment](https://github.com/Homebrew/brew/issues/1670#issuecomment-301273570):

> Most projects have a specific version of Node that they require so as part of each projects bootstrap we install Node to get npm, then use npm to install a thing called "n", an easy NodeJS version switcher.

> We specifically don't want you to pin Node to a particular version because then you don't get any security updates.

**Comment by ghost** [comment]
> on Windows MSVC break ABI from version to version and even between debug and release build.

I cannot emphasize this enough. On windows, everything is built from source using a single compiler.

**Comment by JonathanDCohen** [comment]
Is there a way for us to constantly update the different databases these distros use with our daily release, or a weekly cut, or something of the sort, such that they incrementally keep absl up-to-date?

**Comment by ghost** [comment]
For OSX, the homebrew PR process can be automated (you can just submit a PR bumping the version number every week). Windows is going to use vcpkg, which can be automated as well. 

For Ubuntu, you could create your own ppa (https://askubuntu.com/questions/71510/how-do-i-create-a-ppa), so that it's not bundled with the OS packages, but that's significantly more work.

**Comment by ghost** [comment]
Example for homebrew: https://github.com/Homebrew/homebrew-core/pull/20116

**Comment by a4z** [comment]
even if gcc is not present per default, there is the ubuntu compiler ppa
https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test?field.series_filter=xenial

RHEL has a 'old' base system, but ships newer software and compiles via software collections and compilers come official via devtoolset
https://access.redhat.com/documentation/en/red-hat-developer-toolset

openSUSE has one compiler installed by default, but others are just zypper command far away and easy to install because they are availalbe per default.

and so on...
you have access to a lot of compilers on most distros.

I do not see Abseil as a package in a distribution, so I do not understand why the discussion drifts in this directions.

for distributions Abseil should come via source, vial ppa/software collection/OBS/whatever, of course, because otherwise there will be troubles
but
this has nothing to do that a user should be able to use a make install command of a library/toolset.

A make install is needed for example:
Beside other reasons, when I build a yocto SDK that I have full control about everything, there is only 1 compiler, and when I ship the SDK to users they should find libraries in default directories, include files in default directories, regardless of what build system they use. Abseil should be there and work for them.
Also, in such an environment compiling a file more than once if it is not required or necessary, for each user on each work space, is a wast of time and energy and the opposite of clever engineering.
A source distribution in this scenario is not suitable. Abseil should integrate in the SDK like all other components, and this is done via a make install DESTDIR=/where/ever 
do you agree that make install can be useful for this, @Orphis ?

so I do not understand the discussion about package manager, and the question if you can update database of distros, @JonathanDCohen .  Yes of courser, make a ppa for ubuntu, a software collection for RHEL/CentOS, use OBS for SUSE, whatever distros provide , and you have full control, but as said, for this you want a source distribution. So I see no relation to the make install topic of this issue.

**Comment by junyer** [comment]
I have finally figured out why Abseil can't be distributed: ABI instability not just over time, but due to compiler flags! Bridging `absl::string_view` to `std::string_view`, for example, depends on whether you build with `-std=c++17`. This will trivially break static libraries as well as shared libraries regardless of whether they are in lockstep with the headers. `pkg-config(1)` kind of solves the problem because users can at least discover which flags they should be using, but you can't force anything with it.

**Comment by JonathanDCohen** [comment]
@junyer yes, among other reasons.  You can watch Titus' keynote from cppcon for the broader picture. 

So let's get back to the bug picture.  We want to work with major build tools, and we don't want to be distributed as a binary.  It seems we can accomplish this with install() as a source copier.  Here are my next questions:

1) What does the implementation of this look like? It seems to be pretty straightforward

2) Does this help clients maintain up-to-date forks of Abseil?  As in, can this e meaningfully combined with ExternalProject_Add?  My understanding is that install() is part of that process, so install() is important for Abseil users wanting to maintain an up-to-date copy of Abseil

**Comment by Sarcasm** [comment]
The way I see it, you want to give CMake users a guideline in the line of this:

> Compile Abseil and the other projects in your workspace with the same "CMake toolchain".
>
> `CMAKE_CXX_COMPILER` and `CMAKE_CXX_FLAGS` should matches for all projects in a given workspace.
> Great care should be taken to make sure the C++ standard (`-std=c++xx`) is set in `CMAKE_CXX_FLAGS`,
> so that all projects compile with the same standard.

A "CMake toolchain", is a file that describes the compiler and compile options to use.
- https://cmake.org/cmake/help/v3.9/manual/cmake-toolchains.7.html#variables-and-properties

This makes me realize that using `add_subdirectory()` does not guarantee ABI compatiblitity.
It just guarantees that the user is using the same compiler.
Subdirectories' targets can have their own flags, such as the C++ standard and standard library (it's not rare to see some CMakeLists.txt "smartly" add `-stdlib=libc++`).

---

Regarding the downloading of sources, are you thinking about doing the same as googletest?

- https://github.com/google/googletest/tree/master/googletest#incorporating-into-an-existing-cmake-project

Download a project at configure time?
This looks correct to me, but not really "nice" or clean.
I would qualify this as a hack, I don't see myself use this on a bunch of dependencies.

It can be ok for Abseil workspace but maybe not for everyone integrating Abseil into a workspace.

CMake is not one of this "full stack" build system.

In the Python world, you have:
1. setuptools/setup.py to create a distribution of your project
2. a package manager like pip to deal with dependencies, that reuses some data generated in step 1

CMake is more like setuptools/setup.py, you can create a distribution of your project.
You might abuse it a little to fetch third party libraries but it's not the main purpose of the tool.

IMHO. Abseil should just provide a standard CMake package that can be used with `add_subdirectory()` or `find_package()` from a local install.
Then separately offer a "CMake superbuild" to build the whole Abseil workspace.
Then try to use Abseil with a project like protobuf,
and realize how difficult it is in its current form to integrate into any other projects.
Simplify the CMakeLists.txt, and tell users they need to take care of the dependencies themselves, and use a coherent "CMake toolchain" for the whole workspace.

**Comment by Orphis** [comment]
> This makes me realize that using add_subdirectory() does not guarantee ABI compatiblitity.

It guarantees that it is using the same flags as the parent directory, which we hope are global (either through a toolchain file or good options put at the top level CMakeLists.txt of the top-level project). We can't enforce anything more.

The difference between `add_subdirectory()` and `find_package()` is that we have a chance at least to propagate the current flags that are hopefully global with the former.

Fundamentally, it is the same as people compiling code from a header only library that has multiple implementations selected with a define, and then providing different values in different parts of the code base. You can't prevent that, but you can ensure people use patterns that are less risky.

> Regarding the downloading of sources, are you thinking about doing the same as googletest?

I would consider the first CMake PR as a first pass and more work is needed for proper CMake  integration, probably copying this well written section of the documentation is another one.

The documentation offers basically 2 choices: downloading at configuration time and static files (through submodules or a copy of the files). The first one is certainly a nice "hack", but so is ExternalProject in general. The other ones are totally reasonable for a lot of projects, but of course YMMV.

Downloading at "configure" time is done by many tools anyway, including some CMake package managers. In other language build systems, it can be done in an earlier stage of the integrated build process, so this isn't a foreign concept at all (see cargo for example).

And CMake is perfectly capable of such a workflow, although it's not what people have been traditionally doing.

**Comment by mikezackles** [comment]
What prevents projects consuming Abseil from doing naughty things with their own install targets? Why force a build system on Abseil's consumers instead of facilitating build system agnostic tools that support a [matrix of identifiers](http://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html)? If distros etc. try to start shipping Abseil, why not just explain the issue and ask them to stop? I want to live at HEAD too, but I don't think putting the whole world into CMake is practical, general, or even desirable.

**Comment by isaachier** [comment]
I realize this is an old post, but I plan on adding Abseil to the [Hunter](https://github.com/ruslo/hunter) package manager. I think Hunter really aims to do what Bazel does, but builds the functionality on top of CMake as opposed to reinventing the build system. Hunter is the best when you don't need to worry about compatibility with existing deb packages and the like.

**Comment by JonathanDCohen** [comment]
We're having a team summit this week, and CMake support will come up.  We will be discussing source installation, and also solidifying our plans for periodic stable releases.  So stay tuned!

**Comment by JonathanDCohen** [comment]
Okay everyone.  So a couple things --

We will be tagging long-term-support branches around every 6 months.  There are a few wrinkles internally but this will start soon, and we'll have a more complete announcement when that happens.

As for CMake we're definitely down with source installation -- I left https://github.com/abseil/abseil-cpp/issues/80#issuecomment-373770822 on another issue but I'll sum it up here --

We want to support[ the dance that googletest does](https://github.com/google/googletest/blob/master/googletest/README.md#incorporating-into-an-existing-cmake-project) where you call ExternalProject in a separate cmake process to download abseil source at configure time if a version exists, or otherwise just does add_subdirectory on the found Abseil.

My understanding is that to support this, we need to install(EXPORT) to allow referencing abseil targets, and install(DIRECTORY) to isntal; the actual abseil source.  Explicitly, we are not supporting installation of binaries  or even add_subdirectory with a binary destination.

First, am I correct in what needs to happen on our end?  And second, how do people feel about this approach?

**Comment by Sarcasm** [comment]
To me, it feels like you want the compiler to do its job correctly, that what it compiles and what it uses is not just compatible, but really the same.
But you don't seem to care about using the CMake build system correctly.

Another funny thing, is to promote "live at head", but require ancient time CMake (2.8.12).

You want to force people to have a coherent/correct workspace built with CMake, but CMake is not good to deal with multiple dependencies.
It's good to build one isolated set of related targets: libs, tests and executables.

`add_subdirectory` or External projects are just hacks, is for quick and dirty dependency management, not something to build upon. It's actually CMake antipatterns.
Seeing how it's difficult to have proper CMakeLists.txt, it's an horrible idea to recommend anyone to include someone else CMakeLists.txt, and hope he is a good citizen.
We should just trust the generated dist: lib/, bin/, include/, etc.

Real tools for dependency management, rely on `install` targets, and expertise.

Personnally, I'm thinking that "Configure-time configuration" is possible for Abseil, and it would make it easier to integrate while still being correct.
Unfortunately, easy build system integration is not a primary goal:

- https://groups.google.com/forum/#!topic/abseil-io/uAoewRECskM

I hope the tone is not too harsh.
I'd just like to user both Abseil and CMake correctly.

**Comment by isaachier** [comment]
@sarcasm the only part I disagree with is the way end. CMake does have inherent support for building external projects. That is the basis of one of my favorite projects Hunter (https://github.com/ruslo/hunter). I really believe Hunter is exactly what this project wants, but for some reason or another the motive to keep using classic CMake is too powerful. For the record, Hunter uses backwards compatible extensions so the install works even without downloading. I highly recommend building something with it. I guarantee you'll like it.

**Comment by gennadiycivil** [comment]
@Sarcasm , re: CMake 2.8.12 - good probability that Abseil will require much newer CMake soon.

**Comment by JonathanDCohen** [comment]
@Sarcasm I'm a little confused at your response.  I'm basically proposing doing https://github.com/google/googletest/blob/master/googletest/CMakeLists.txt#L114 for Abseil, allowing it to be used in the same way as googletest.  I'm not sure what I'm missing.

**Comment by isaachier** [comment]
I hate to use reddit as a source, but this is a pretty good summary of the issue:

"I've used both pretty extensively at work, and use Bazel for my hobby projects. They're great to work with, so long as you buy into the monorepo pitch. Hermetic, reproducible builds...
The big problem is that they're a little bit all or nothing -- whatever third party project you want to depend on probably assumes an autotools deployment strategy: you download, make, and install your dependencies in a way your project doesn't need to understand, and then find them in whatever directory you installed them to within your project. There's a lot of problems with this way of doing things, but the reality is that you can't rewrite the world, and will end up either making a build target that just globs together all your dependency's sources, or else invokes its build system from within yours."

https://www.reddit.com/r/cpp/comments/6euc7b/_/die6g1y?context=1000

**Comment by Sarcasm** [comment]
@JonathanDCohen 

The GoogleTest CMake snippet you linked to, installs the binary targets:

```cmake
cxx_library(gtest "${cxx_strict}" src/gtest-all.cc)
cxx_library(gtest_main "${cxx_strict}" src/gtest_main.cc)
target_link_libraries(gtest_main gtest)

...

install(TARGETS gtest gtest_main
  RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
  ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
  LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}")
install(DIRECTORY "${gtest_SOURCE_DIR}/include/gtest"
  DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
```

https://github.com/google/googletest/blob/dccc2d67547a5a3a97e4f211f39df931c6fbd5d5/googletest/CMakeLists.txt#L114

If that's what you propose, then I'm all for it, and I'm sorry for the misunderstanding.

Sadly, I don't think that's what you proposed by rereading this (emphasis mine):

> My understanding is that to support this, we need to install(EXPORT) to allow referencing abseil targets, and install(DIRECTORY) to isntal; the actual abseil source.
> Explicitly, **we are not supporting installation of binaries** or even add_subdirectory with a binary destination.
-- https://github.com/abseil/abseil-cpp/issues/38#issuecomment-373776032


So, if your actual plan, is to do like GoogleTest except that you don't install the binary targets, then I maintain my comment, and additionally the comments I made here, against `add_subdirectory` and `ExternalProject`, also applies to GoogleTest.

**Comment by JonathanDCohen** [comment]
I'm wondering if there isn't a miscommunication here, after talking with some other people internally.

Is the concern that you'd like to do something like `apt-get install abseil` and then find_package(Abseil)?

**Comment by coryan** [comment]
While I do not expect the Abseil developers to provide packages for every platform out there (`apt-get install abseil`, and `vcpkg install abseil`, and `brew install abseil`, and `dnf install abseil`, etc. etc.) I would like them to make it easy for package maintainers to create such packages.  Part of making that easy is that `make install` should work and install the necessary headers, libraries, pkg-config files, and CMake config files, in the (more or less) standard places.

Otherwise each package maintainer has to reimplement the `install` target themselves, and mistakes are more likely to happen.

**Comment by Sarcasm** [comment]
I firmly believe the clean way to use packages with CMake is to use `find_package`.
I'm not interested in system-wide installation of Abseil, I install locally and use `-DCMAKE_PREFIX_PATH=<path/to/local/abseil/install>`.
I don't want my project to rely on `add_subdirectory`, and to break at any unexpected change from the Abseil CMakeLists.txt after an upgrade.

I'd like the `install` rule because a lots of existing tools can relies on it correctly:
- https://github.com/abseil/abseil-cpp/issues/38#issuecomment-338779633

Sure, these tools can be misused, but they can also be used perfectly well, and when a package like Abseil just add difficulties because of "fear of misuse", it's seems wrong.
I prefer things to be simple. You declare targets, you install them. I find_package Abseil, use the exported targets.

The CMake documentation documents how dependencies are to be used:
- https://cmake.org/cmake/help/v3.11/manual/cmake-packages.7.html#using-packages

I don't want the Abseil team to teach me how I should use CMake in a different way than recommended by the CMake documentation itself. Either you adopt the build system and its quirks, or you give up and just say "use Bazel".

Adopting CMake and propagating the unconventional usage is not doing any good IMHO.


Disclaimer: I understand the complications of system wide installation, the ABI issues you can have as a system installation cannot guarantee every use of the library will use the correct settings. However, for projects at my companies, we don't do system wide installations, we have "workspaces" where we build the library, and make sure we have a set of correct build settings. We rely on autotools projects, CMake projects, Boost.Build, ...All these tools provide installation targets and we rely on them.

**Comment by JonathanDCohen** [comment]
Okay great, I think I understand the issues here.  Let me make my goals clear from my end, and hopefully we can move forward with this:

What I'm aiming to do *right now* is support live-at-head use of Abseil with CMake.  This necessarily involves the configuration-time dance that googletest does, or else users have a superbuild and use ExternalPRoject

In the near future, abseil will start to cut long-term support branches about every 6 months, likely with versioned inline namespaces.  I and other members of the team are pushing that we should work with package manager maintainers such that all of their packages can, as much as possible, use the same long term support version of Abseil.

Now to your point, @Sarcasm 

> I'm not interested in system-wide installation of Abseil, I install locally and use -DCMAKE_PREFIX_PATH=<path/to/local/abseil/install>.

I follow you here, but allow me to ask a stupid question -- how are you installing Abseil in this situation without a package manager?

@coryan

> make install should work and install the necessary headers, libraries, pkg-config files, and CMake config files, in the (more or less) standard places.

For headers, we may be able to cobble together some kind of "includes" subdirectory of an install destination.  What we probably can't do is actually structure abseil's code in this way -- it would be way too much work to restructure our code at this point.  

As for libraries, abseil is actively shedding our cctz dependency, at which point we will have no non-test dependencies.  This seems to me like then there are no libraries to install.   

I have to admit I don't know what pkg-config is, I'll have to research that.  

For CMake config files, what I got out of our training with Kitware is largely that absl/base/config.h does what would otherwise be done with a cmake config file, unless I'm misunderstanding you.

Again, apologies if I'm being super dense here.  I'm coming at this with the best intentions and learning on the fly.

**Comment by isaachier** [comment]
Thank you @JonathanDCohen for your dedication here. First, I'm not sure how `absl/base/config.h` does anything like a CMake config. I think you might be confusing a compiler config with a CMake config. So let me give a brief explanation.

A CMake config (and pkg-config for that matter) solve a pretty basic problem. If I have library B that depends on library A, how do I determine the necessary header directories, libraries, and flags to compile with? CMake developed two solutions to the problem.

1. Have each package write a `FindA.cmake` file dedicated to resolving all requirements for linking/compiling against A.
2. Have a global configuration installed in `<prefix>/lib/cmake/AConfig.cmake` that resolves these requirements for any dependency.

The problem with the first solution should be clear. The package requirements should be dictated by the package itself. The burden of determining required flags should not be placed on the dependent library. Unfortunately, most CMake projects to this day still rely almost entirely on this solution.

The second solution is much more sane, and is compounded by the fact that CMake can do all the work for you. If library A has a CMake build, CMake can easily create a config for consumption by library B (see https://cmake.org/cmake/help/v3.1/module/CMakePackageConfigHelpers.html for details). The primary issue here is the inability to apply this to non-CMake projects.

Now that we have that sorted out, you can see why I strongly suggest you take the second, more sensible approach.

**Comment by coryan** [comment]
Thank you for clarifying the short-term goals.  You may want to open a separate bug or something called "Partial improvements to CMake support" or show folks a roadmap (without dates) where you explain how this fits into the plan to have releases with versioned namespaces and so forth.

> > make install should work and install the necessary headers, libraries, pkg-config files, 
> > and CMake config files, in the (more or less) standard places.

> For headers, we may be able to cobble together some kind of "includes" subdirectory of an install
> destination. What we probably can't do is actually structure abseil's code in this way -- it would be way
> too much work to restructure our code at this point.

Agree, `/usr/include/absl/*.h` (and deeper subdirectories as needed), is what I clumsily trying to say.  Or more generally `${prefix}/include/absl/**/*.h` in case the users decide to install in `/opt`, or `/usr/local`, or some other directory.  What would be surprising and I dare say unacceptable, would be to install in `${prefix}/share/doc` or `${prefix}/src`.

> As for libraries, abseil is actively shedding our cctz dependency, at which point we will have no
> non-test dependencies. This seems to me like then there are no libraries to install.

I was skipping steps and thinking about your longer term goals.  In that case you do create `libabsl_base.a`, right?  Those should get installed to `${prefix}/lib` (typically, it is different on Windows, and different on some flavors of Linux, but that is Too Much Detail).

**Comment by nolange** [comment]
> I follow you here, but allow me to ask a stupid question -- how are you installing Abseil in this situation without a package manager?

In a way - thats the whole point why we nag about an install target. Normally you would do this:

```bash
# thats your workspace with all your libraries in whatever version you want.
MYBUILD_DIR=<someplace>
# git clone / unpack your sources to /tmp/abseil/src

cd /tmp/abseil/build
cmake -DCMAKE_PREFIX_PATH=$MYBUILD_DIR ../src
make
DESTDIR=$MYBUILD_DIR make install

# repeat for all dependencies

cd <my_project>/build
cmake -DCMAKE_PREFIX_PATH=$MYBUILD_DIR ../src
make
DESTDIR=/tmp/my_project_package make install
# now I can pack up /tmp/my_project_package, and unpack this to its final destination
```

it should not matter whether you use a package manager or build from source, atleast for your project files. I don't know how the part before can be solved easily in an automated foolproof flexible configuration (I am leaning toward an extra CMakeList.txt thats independent from the my_project build).

but the key point is that you have an directory that's your build environment and you dont want to leak out details from individual builds, and you dont want to burden the users of your library by having to manually extract all files and configuration. You do so by providing an install target which can be redirected (CMake mostly does this for you, as do Automake and everything else).
After that you can delete the build directory (`/tmp/abseil/`).

Or you could use some Projects like Buildroot / Yocto to create this build environment for you. At any rate you will need a way to install your libs/headers/source and a way to use them in depended projects.

**Comment by Sarcasm** [comment]
> Now to your point, @Sarcasm
> 
> > I'm not interested in system-wide installation of Abseil, I install locally and use
> > `-DCMAKE_PREFIX_PATH=<path/to/local/abseil/install>`.
> 
> I follow you here, but allow me to ask a stupid question -- how are you installing Abseil in this situation without a package manager?

If I want to try a new library, like Abseil, I proceed as follow:

First, I have an existing workspace with just my app:

```
$ pwd
/my/workspace
$ ls
my_app/
```

I will first get the Abseil cctz dependency:

```
mkdir third_party
cd third_party
git clone https://github.com/google/cctz.git
mkdir cctz/build
cd cctz/build
cmake -G Ninja -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=/my/workspace/third_party-install ..
ninja -v        # check compile options are ok (e.g. -std=c++11), adjust if necessary
ninja install   # (1) sadly this does not work yet because they want to do like Abseil: https://github.com/google/cctz/pull/68#issuecomment-364585683
```

Then Abseil (notice how this could be the same steps as cctz,
that's the advantage of using conventional CMake):

```
cd /my/workspace/third_party
git clone https://github.com/abseil/abseil-cpp.git
mkdir abseil-cpp/build
cd abseil-cpp/build
# (2) this cmake call does't work, because Abseil does not consume cctz with `find_package`
cmake -G Ninja -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=/my/workspace/third_party-install ..
ninja -v         # check compile options are ok, adjust if necessary
ninja install    # (3) one can always hope!
```

(1), (2) and (3) are all the reasons why departing from standard CMake is really painful,
none of them work right now because of this issue.

(2) the following patch was sufficient:

```diff
+find_package(cctz REQUIRED)
+set(ABSL_CCTZ_TARGET cctz::cctz)

- if(NOT ABSL_CCTZ_TARGET)
-   set(ABSL_CCTZ_TARGET cctz)
- endif()
```

Now I have the following workspace:

```
$ pwd
/my/workspace
$ ls
my_app/
third_party/
third_party-install/
$ ls third_party-install/*
third_party-install/include:
absl   cctz

third_party-install/lib:
cmake  libabsl_base.a  ... libcctz.a
```

And to test it in `my_app/`, I can do the following steps:

Edit my_app/CMakeLists.txt:

```diff
+find_package(absl REQUIRED)
+target_link_libraries(my_app absl::base)
```

Reconfigure CMake to specify where to find the dependencies:

    cmake -DCMAKE_PREFIX_PATH=/my/workspace/third_party-install .



What's really important here is that you build on existing knowledge/practices.
Even beyond CMake, this method works mostly the same with autotools, and probably things like Meson.

You can see this is already a bit painful this way, not providing the install targets makes it even more so.


As @nolange says, tools like Buildroot/Yocto really works with this kind of workflow.
Just an interesting example, see this Yocto best practices for CMake, the second item is:

> 2. do all install operations in CMake if possible (vs do_install in the OE recipe)
> -- http://bec-systems.com/site/1128/best-practices-for-using-cmake-in-openembeddedyocto



> @JonathanDCohen says:
>
> I have to admit I don't know what pkg-config is, I'll have to research that.

Yes, please read about it.
Some thinks a successful package manager for C++ needs something like pkg-config.
Autotools, CMake, Meson and probably others, support pkg-config.
CMake supports it but it also has it's own specific config file format.

> For CMake config files, what I got out of our training with Kitware is largely that absl/base/config.h does what would otherwise be done with a cmake config file, unless I'm misunderstanding you.

For me a CMake config file is xxx-config.cmake, is not the same as absl/base/config.

- xxx-config.cmake is to declare your target and usage requirements: include path, library to link, dependencies, ...
- absl/base/config.h  is a configuration done at "compile time": I do believe, configuration should be done at "configuration time": https://groups.google.com/forum/#!topic/abseil-io/uAoewRECskM
- CMake support config.h generation with `configure_file`: https://cmake.org/cmake/help/latest/command/configure_file.html


If you can, you should say hi to the `#cmake` channel on cpplang' Slack: https://cpplang.now.sh/
People there have really good feedbacks on using CMake correctly and using modern idioms.

**Comment by Quincunx271** [comment]
Could the FindFoo.cmake or foo-config.cmake be used to do a `git pull` then build/test if necessary? That would allow for the nice `find_package(absl)` syntax, while also keeping abseil up to date. Seems like abuse of those files, though.

**Comment by isaachier** [comment]
@Quincunx271 I do not believe so. Those files are always used to find *precompiled* content, not source files.

**Comment by nolange** [comment]
downloading anything during the configure stage is a anti-pattern to me, but no one forces the libraries that are found with `find_package` to be precompiled.

There are object libraries in CMake which could be used for that.
```
find_package(Abseil)

# add abseil source files
add_executable(my_project
  $<TARGET_OBJECTS:Abseil::Sources> 
)
# add usage requirements: include path, macros, depended libraries...
target_link_libraries(my_project
  Abseil::Common
)
```

**Comment by isaachier** [comment]
Well as I've said before, if that is what you are looking for, use ExternalProject or Hunter (https://github.com/ruslo/hunter).

**Comment by JonathanDCohen** [comment]
This is now tracked in #111

**Comment by lilianmoraru** [comment]
Finally. I can attempt now to push Abseil at work :)
